// graphql-zeus-core/lib/TreeToTS/functions/apiSubscription/graphql-ws.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.apiSubscription = void 0;
const graphql_ws_1 = require("graphql-ws");
const apiSubscription = (options) => {
    var _a;
    const client = (0, graphql_ws_1.createClient)({
        url: String(options[0]),
        connectionParams: Object.fromEntries(new Headers((_a = options[1]) === null || _a === void 0 ? void 0 : _a.headers).entries()),
    });
    const ws = new Proxy({
        close: () => client.dispose(),
    }, {
        get(target, key) {
            if (key === 'close')
                return target.close;
            throw new Error(`Unimplemented property '${String(key)}', only 'close()' is available.`);
        },
    });
    return (query) => {
        let onMessage;
        let onError;
        let onClose;
        client.subscribe({ query }, {
            next({ data }) {
                onMessage && onMessage(data);
            },
            error(error) {
                onError && onError(error);
            },
            complete() {
                onClose && onClose();
            },
        });
        return {
            ws,
            on(listener) {
                onMessage = listener;
            },
            error(listener) {
                onError = listener;
            },
            open(listener) {
                client.on('opened', listener);
            },
            off(listener) {
                onClose = listener;
            },
        };
    };
};
exports.apiSubscription = apiSubscription;
