// graphql-zeus-core/lib/TreeToTS/functions/new/mocks.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HEADERS = exports.HOST = exports.apiFetch = exports.apiSubscription = exports.Ops = exports.ReturnTypes = exports.AllTypesProps = void 0;
exports.AllTypesProps = {
    Query: {
        cardByStatus: {
            status: 'Status',
        },
        cards: {
            attack: 'TypeOfAttack',
        },
    },
    Mutation: {
        createCard: {
            card: 'CreateCard',
        },
    },
    Card: {
        attack: {
            by: 'Status',
        },
    },
    CreateCard: {
        status: 'Status',
        settings: 'JSON',
    },
    Status: 'enum',
    TypeOfAttack: 'enum',
    JSON: 'scalar.JSON',
};
exports.ReturnTypes = {
    Query: {
        cardByStatus: 'Card',
        drawCard: 'Card',
        cards: 'Card',
    },
    Mutation: {
        getCardAndPop: 'Card',
    },
    Card: {
        attack: 'Card',
        info: 'JSON',
    },
    JSON: 'scalar.JSON',
};
exports.Ops = {
    query: 'Query',
    mutation: 'Mutation',
};
const apiSubscription = (opts) => ((q) => 1);
exports.apiSubscription = apiSubscription;
const apiFetch = (options) => ((q, vars) => 1);
exports.apiFetch = apiFetch;
exports.HOST = 'https://faker.graphqleditor.com/a-team/olympus/graphql';
exports.HEADERS = {};
